File: access_control.java
Here's an example of a simple access control mechanism using Java. This program will create a `User` class with a username and role, and a `AccessControl` class to manage access to certain operations based on the role of the user.

```java
// User class to hold user information
class User {
    private String username;
    private String role;

    public User(String username, String role) {
        this.username = username;
        this.role = role;
    }

    public String getUsername() {
        return username;
    }

    public String getRole() {
        return role;
    }
}

// AccessControl class to manage access to certain operations
class AccessControl {
    public void grantAccess(User user, String operation) {
        // Check user role before granting access
        if (user.getRole().equals("admin")) {
            System.out.println("Access granted to " + user.getUsername() + " for operation " + operation);
        } else {
            System.out.println("Access denied to " + user.getUsername() + " for operation " + operation);
        }
    }
}

// Main class to test the access control mechanism
public class Main {
    public static void main(String[] args) {
        User admin = new User("admin", "admin");
        User user = new User("user", "user");

        AccessControl ac = new AccessControl();
        ac.grantAccess(admin, "delete");
        ac.grantAccess(user, "delete");
    }
}
```

In this program, only users with the role "admin" are allowed to perform the "delete" operation. If a user with a different role tries to perform this operation, the access control mechanism will deny access. This is a simple example, but you can extend this concept to create more complex access control mechanisms.

Note: In a real-world scenario, you might want to use more sophisticated methods for access control, such as Access Control Lists (ACLs) or Role-Based Access Control (RBAC), and you would likely integrate with a database or other persistent storage to manage users and roles.

File: access_control.go
Here's a simple Go program that implements an access control system. This program uses a map to store users and their corresponding roles. The program has functions to add users, assign roles, and check if a user has a certain role. 

```go
package main

import (
	"fmt"
)

// User struct represents a user in the system
type User struct {
	Name string
	Role string
}

// AccessControl struct represents the access control system
type AccessControl struct {
	Users map[string]User // map of users with their name as the key
}

// NewAccessControl creates a new access control system
func NewAccessControl() *AccessControl {
	return &AccessControl{Users: make(map[string]User)}
}

// AddUser adds a new user to the access control system
func (ac *AccessControl) AddUser(name string, role string) {
	ac.Users[name] = User{Name: name, Role: role}
}

// AssignRole assigns a role to a user
func (ac *AccessControl) AssignRole(name string, role string) {
	user, exists := ac.Users[name]
	if exists {
		user.Role = role
		ac.Users[name] = user
	}
}

// HasRole checks if a user has a certain role
func (ac *AccessControl) HasRole(name string, role string) bool {
	user, exists := ac.Users[name]
	if exists {
		return user.Role == role
	}
	return false
}

func main() {
	ac := NewAccessControl()

	ac.AddUser("Alice", "admin")
	ac.AddUser("Bob", "user")

	fmt.Println(ac.HasRole("Alice", "admin")) // Output: true
	fmt.Println(ac.HasRole("Bob", "admin"))   // Output: false

	ac.AssignRole("Bob", "admin")

	fmt.Println(ac.HasRole("Bob", "admin")) // Output: true
}
```

This program is a simple representation of an access control system and does not include any kind of authentication or password protection for users. In a real-world application, you would need to implement these features as well as others such as role hierarchies, permissions, etc.